File: 백준/Gold/28707. 배열 정렬/배열 정렬.py
import sys
import heapq

N = int(sys.stdin.readline())
A = list(map(int,sys.stdin.readline().split()))
M = int(sys.stdin.readline())
cost = []
for _ in range(M) :
    l, r ,c = map(int,sys.stdin.readline().split())
    cost.append((l-1,r-1,c))

answer = sorted(A)
for i in range(N) :
    if answer[i] == 10 :
        answer[i] = 0
answer = int(''.join(map(str,answer)))

for i in range(N) :
    if A[i] == 10 :
        A[i] = 0
A = int(''.join(map(str,A)))

queue = []
dp = [float('inf')] * (10**N)
dp[A] = 0
heapq.heappush(queue, (0,A))

while queue :
    now_cost, now_node = heapq.heappop(queue)

    if now_cost > dp[now_node] :
        continue

    for l, r, c in cost :
        next_node = list(str(now_node))
        if len(next_node) != N :
            n_node = ['0'] * N
            for i in range(len(next_node)-1,-1,-1) :
                n_node[i+(N-len(next_node))] = next_node[i]
            next_node = n_node

        tmp = next_node[l]
        next_node[l] = next_node[r]
        next_node[r] = tmp
        next_node = int(''.join(map(str,next_node)))
        dis = now_cost + c
        if dis < dp[next_node] :
            dp[next_node] = dis
            heapq.heappush(queue, (dis,next_node))

if dp[answer] == float('inf') :
    print(-1)
else :
    print(dp[answer])

File: 백준/Gold/28707. 배열 정렬/배열 정렬.java
import java.io.*;
import java.util.*;

class Cost {
    int l, r, c;
    public Cost(int l, int r, int c) {
        this.l = l;
        this.r = r;
        this.c = c;
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        // 입력 처리
        int N = Integer.parseInt(br.readLine());
        int[] A = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int M = Integer.parseInt(br.readLine());
        List<Cost> costs = new ArrayList<>();
        for (int i = 0; i < M; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int l = Integer.parseInt(st.nextToken()) - 1;
            int r = Integer.parseInt(st.nextToken()) - 1;
            int c = Integer.parseInt(st.nextToken());
            costs.add(new Cost(l, r, c));
        }

        // 목표 상태 계산
        int[] sortedA = Arrays.copyOf(A, N);
        Arrays.sort(sortedA);
        StringBuilder targetBuilder = new StringBuilder();
        for (int num : sortedA) {
            targetBuilder.append(num == 10 ? 0 : num);
        }
        String target = targetBuilder.toString();

        // 초기 상태 계산
        StringBuilder startBuilder = new StringBuilder();
        for (int num : A) {
            startBuilder.append(num == 10 ? 0 : num);
        }
        String start = startBuilder.toString();

        // 다익스트라 준비
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(o -> o[0]));
        Map<String, Integer> dp = new HashMap<>();
        dp.put(start, 0);
        pq.add(new int[]{0, Integer.parseInt(start)});

        while (!pq.isEmpty()) {
            int[] current = pq.poll();
            int nowCost = current[0];
            String nowNode = String.format("%0" + N + "d", current[1]); // 자릿수 맞추기

            if (nowCost > dp.getOrDefault(nowNode, Integer.MAX_VALUE)) {
                continue;
            }

            for (Cost cost : costs) {
                char[] nextNode = nowNode.toCharArray();
                char temp = nextNode[cost.l];
                nextNode[cost.l] = nextNode[cost.r];
                nextNode[cost.r] = temp;

                String nextNodeStr = new String(nextNode);
                int newCost = nowCost + cost.c;

                if (newCost < dp.getOrDefault(nextNodeStr, Integer.MAX_VALUE)) {
                    dp.put(nextNodeStr, newCost);
                    pq.add(new int[]{newCost, Integer.parseInt(nextNodeStr)});
                }
            }
        }

        // 결과 출력
        if (dp.getOrDefault(target, Integer.MAX_VALUE) == Integer.MAX_VALUE) {
            System.out.println(-1);
        } else {
            System.out.println(dp.get(target));
        }
    }
}
